meilisearch-schema/src/position_map.rs

```diff
@@ -0,0 +1,161 @@
+use std::collections::BTreeMap;
+
+use crate::{FieldId, IndexedPos};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct PositionMap {
+    pos_to_field: Vec<FieldId>,
+    field_to_pos: BTreeMap<FieldId, IndexedPos>,
+}
+
+impl PositionMap {
+    /// insert `id` at the specified `position` updating the other position if a shit if caused by
+    /// the operation. If `id` is already present in the position map, it is moved to the requested
+    /// `position`, potentially causing shifts.
+    pub fn insert(&mut self, id: FieldId, position: IndexedPos) -> IndexedPos {
+        let mut upos = position.0 as usize;
+        let mut must_rebuild_map = false;
+
+        if let Some(old_pos) = self.field_to_pos.get(&id) {
+            let uold_pos = old_pos.0 as usize;
+            self.pos_to_field.remove(uold_pos);
+            must_rebuild_map = true;
+        }
+
+        if upos < self.pos_to_field.len() {
+            self.pos_to_field.insert(upos, id);
+            must_rebuild_map = true;
+        } else {
+            upos = self.pos_to_field.len();
+            self.pos_to_field.push(id);
+        }
+
+        // we only need to update all the positions if there have been a shift a some point. In
+        // most cases we only did a push, so we don't need to rebuild the `field_to_pos` map.
+        if must_rebuild_map {
+            self.field_to_pos.clear();
+            self.field_to_pos.extend(
+                self.pos_to_field
+                .iter()
+                .enumerate()
+                .map(|(p, f)| (*f, IndexedPos(p as u16))),
+            );
+        } else {
+            self.field_to_pos.insert(id, IndexedPos(upos as u16));
+        }
+        IndexedPos(upos as u16)
+    }
+
+    /// Pushes `id` in last position
+    pub fn push(&mut self, id: FieldId) -> IndexedPos {
+        let pos = self.len();
+        self.insert(id, IndexedPos(pos as u16))
+    }
+
+    pub fn len(&self) -> usize {
+        self.pos_to_field.len()
+    }
+
+    pub fn field_to_pos(&self, id: FieldId) -> Option<IndexedPos> {
+        self.field_to_pos.get(&id).cloned()
+    }
+
+    pub fn pos_to_field(&self, pos: IndexedPos) -> Option<FieldId> {
+        let pos = pos.0 as usize;
+        self.pos_to_field.get(pos).cloned()
+    }
+
+    pub fn field_pos(&self) -> impl Iterator<Item = (FieldId, IndexedPos)> + '_ {
+        self.pos_to_field
+            .iter()
+            .enumerate()
+            .map(|(i, f)| (*f, IndexedPos(i as u16)))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_default() {
+        assert_eq!(
+            format!("{:?}", PositionMap::default()),
+            r##"PositionMap { pos_to_field: [], field_to_pos: {} }"##
```

Comment:
I am not quite sure why you use the raw string syntax (`r##`) here and in nexts asserts?

Reply:
I initially used this [neat macro](https://github.com/aaronabramov/k9#inline-snapshots) for testing, that generated thoses strings and inlined them. But as noted by @LegendreM, this is insecure, so I just replaced the macro with `assert_eq`.

Raw string are not *necessary*, but this is how the codegen put them, and I saw no reason to remove it.